pkg/cafs/cafs.go

```diff
@@ -252,3 +259,124 @@ func IsRootKey(fs storage.Store, key Key, leafSize uint32) bool {
 	}
 	return len(keys) > 0
 }
+
+func AddCaching(fsImpl Fs, cacheStore storage.Store, cafsKey2StoreKey func(Key) (string, error)) Fs {
+	return &cachedFs{
+		fsImpl:           fsImpl,
+		cacheStore:       cacheStore,
+		cafsKey2StoreKey: cafsKey2StoreKey,
+	}
+}
+
+type cachedFs struct {
+	fsImpl           Fs
+	cacheStore       storage.Store
+	cafsKey2StoreKey func(Key) (string, error)
+}
+
+func (fs *cachedFs) Put(ctx context.Context, src io.Reader) (int64, Key, []byte, bool, error) {
+	bytesSlice, err := ioutil.ReadAll(src)
+	if err != nil {
+		return 0, Key{}, nil, false, err
+	}
+	var bytesBuf bytes.Buffer
+	bytesBuf.Write(bytesSlice)
+	written, key, keys, found, err := fs.fsImpl.Put(ctx, &bytesBuf)
+	if err != nil {
+		return 0, Key{}, nil, false, err
+	}
+	storeKey, err := fs.cafsKey2StoreKey(key)
+	if err != nil {
+		return 0, Key{}, nil, false, err
+	}
+	err = fs.cacheStore.Put(ctx, storeKey, &bytesBuf, storage.IfNotPresent)
+	if err != nil {
+		return 0, Key{}, nil, false, err
+	}
+	return written, key, keys, found, err
+}
+
+// ??? thread-safety
```

Comment:
For thread safety, the cacheStore can write to a file and move into place. These are CAFS files so safe in terms of over writes.

Reply:
i've added this to #158 and will open a separate request to this effect

Reply:
i made a `storage.Store` implementation that relies on move of the underlying `afero.Fs` to provide atomicity and used it in `bundle_mount.go` as suggested here.

considering move as the atomic part of filesystem operations considered insightful.